### DISCUSSION TITLE
Questions using Silero with Microsoft.ML.OnnxRuntime

### DISCUSSION THREAD

**Discussion by kkppstudios** [snakers4/silero-vad#216] [Q&A] [answered]
Hi. I am playing around with silero-vad and I think I've had some success, but I have some questions about the input and output.

**Background:**
I am developing an Android app with Xamarin Forms. I am using the nuget package Microsoft.ML.OnnxRuntime. I am streaming audio from the device's microphone and trying to process it in real-time. The audio is 16kHz, PCM-16bit, one channel. I am converting the audio bytes into a normalized (-1,1) float array, then batching it in batch sizes of 1536 frames/samples. So far so good. I can create a OnnxRuntime InferenceSession using the .onnx model. I can send the audio data in as input and I am seeing some output.

**Questions:**
After enough searching and playing around, I found that the inputs are "input", "h0", and "c0". What is h0 and c0? To satisfy the model, I am sending in an empty float[] with dimensions 2,1,64 for each of these inputs. I have no idea what they do and if I am setting them up correctly.
As for the output, I am seeing two floats. My understanding was that this model would return one float that represented the probability of speech. So why are there two floats and what do each of them mean? There is also a "hn" and "cn" being output, which I am again ignoring for now.

Any insight into this would be really helpful. I think I am getting close, I just need to understanding these inputs/outputs.

**Comment by snakers4** [comment, marked as answer]
Hi,

> After enough searching and playing around, I found that the inputs are "input", "h0", and "c0". What is h0 and c0?

You can just follow this wrapper logic in your code:

https://github.com/snakers4/silero-vad/blob/7c671a75c2f82830a550ab4a182e86d3a2609d70/utils_vad.py#L10-L52
`input` is just input samples, `h0` and `c0` are just internal model states that are kept between iterations. We did not find an easier way to do so with ONNX.

> My understanding was that this model would return one float that represented the probability of speech.

This line should help:

https://github.com/snakers4/silero-vad/blob/7c671a75c2f82830a550ab4a182e86d3a2609d70/utils_vad.py#L47-L50

**Reply by shrestha-pranita** [reply]
@kkppstudios Hi!! I have been trying to create an  empty float[] with dimensions 2,1,64 to pass as h0 and c0, however, I keep getting byteLength error. Could you post how you created these?

**Reply by kkppstudios** [reply]
In my class I have the following:

```
private readonly float[] h0Value = new float[128];
private readonly float[] c0Value = new float[128];

private readonly NamedOnnxValue h0NamedOnnxValue = null;
private readonly NamedOnnxValue c0NamedOnnxValue = null;
```

In the class constructor:

```
var h0Input = new DenseTensor<float>(this.h0Value, new[]
{
    2, 1, 64
});

var c0Input = new DenseTensor<float>(this.c0Value, new[]
{
    2, 1, 64
});

this.h0NamedOnnxValue = NamedOnnxValue.CreateFromTensor("h0", h0Input);
this.c0NamedOnnxValue = NamedOnnxValue.CreateFromTensor("c0", c0Input);

```

Then in a function within the class, I call session.Run:

```
using var results = this.session.Run(new List<NamedOnnxValue>
{
    NamedOnnxValue.CreateFromTensor("input", input),
    this.h0NamedOnnxValue,
    this.c0NamedOnnxValue
});
```

Note, my use case is streaming audio.

**Reply by shrestha-pranita** [reply]
Thanks for the help. For the feeds's input are you passing the input in the shape of (1, .....)?? I keep getting the following error:

E:onnxruntime:, sequential_executor.cc:368 Execute] Non-zero status code returned while running Conv node. Name:'Conv_81' Status Message: Invalid input shape: {1}
